Refactor and test alternate live data loaders

TylerLubeck opened the pull request:
This is an alternate version of #674 .

In this version, I've pulled some things apart so I can more thoroughly test them. I split them up because this is a more complex change, but I prefer this one over the other.

TylerLubeck committed:
Accept json data that is not a dictionary

TylerLubeck committed:
Refactor and test alternate live data loaders

nsheaps commented on baseplate/sidecars/live_data_watcher.py:
```diff
@@ -35,78 +31,23 @@
 
 
+class LoaderType(Enum):
+    RAW = 'RAW'
```
🔕  "Raw" to me implies some parsing of the data (like mime type), "default"/"zookeeper"/"direct" might be better terminology.

KTAtkinson commented on baseplate/sidecars/live_data_watcher.py:
```diff
@@ -128,25 +73,67 @@ def on_change(self, data: Optional[bytes], _znode_stat: ZnodeStat) -> None:
+    )
+
+    try:
+        s3_object = s3_client.get_object(**s3_kwargs)
```
What error would be raised here if the object or bucket didn't exist?

TylerLubeck replied:
[This](https://stackoverflow.com/questions/33842944/check-if-a-key-exists-in-a-bucket-in-s3-using-boto3) is an admittedly old reference, but it appears to be a `ClientError`

KTAtkinson commented on baseplate/sidecars/live_data_watcher.py:
```diff
@@ -128,25 +73,67 @@ def on_change(self, data: Optional[bytes], _znode_stat: ZnodeStat) -> None:
+
+    s3_client = boto3.client(
+        "s3",
+        region_name=region_name,
```
What happens here if the region is invalid?

TylerLubeck replied:
`EndpointConnectionError`, I'm adding coverage for it now

nsheaps commented on baseplate/sidecars/live_data_watcher.py:
```diff
@@ -128,25 +73,67 @@ def on_change(self, data: Optional[bytes], _znode_stat: ZnodeStat) -> None:
+    try:
+        loader_type = json_data["live_data_watcher_load_type"]
+    except (json.decoder.JSONDecodeError, KeyError, AttributeError, TypeError):
+        logger.debug("Data is not a JSON object like {\"live_data_watcher_load_type\":\"load_type\"}. Loading as RAW")
```
```suggestion
        logger.debug("Expected dict (JSON object) but got %s. Loading as RAW" % str(type(json_data)) )
```

nsheaps commented on baseplate/sidecars/live_data_watcher.py:
```diff
@@ -128,25 +73,67 @@ def on_change(self, data: Optional[bytes], _znode_stat: ZnodeStat) -> None:
+            error.response["Error"]["Message"],
+        )
+    except ValueError as error:
+        logger.exception(
```
🔕  I think technically it's more pythonic to just raise an exception than to log it and return `None`. Doesn't make much of a difference to me.

TylerLubeck replied:
If we raise an exception then we either have to handle it elsewhere or have the data watcher fail and exit.

IMO returning None means "we got no data" and that's a valid thing to write out as data.

I can be convinced otherwise pretty easily

nsheaps replied:
"No data" is valid aka empty file (though I'd expect an empty file locally too)

We tried to get this and got an error (4xx, 5xx), I would not expect that to dump an empty file

nsheaps replied:
If it were to return None, you'd just end up getting an exception when trying to write the file instead:

https://github.com/reddit/baseplate.py/pull/676/files#diff-f08bba0851fd42e4ae2a8d9fb6b983e377e46214ca4fbb35a3738f4d97665630R76

![Screen Shot 2022-05-20 at 2 28 50 PM](https://user-images.githubusercontent.com/1282393/169590519-2de3bcf1-6ef7-4280-9e06-00beb4163322.png)

TylerLubeck replied:
hm ok I can see it. Another edge case to consider then:

Let's say the s3 object was created, we had watchers fetch it, then we deleted it from s3 and fetchers ran again. Now they've got an "object doesn't exist error" (or whatever the error is). Do we delete the existing local copy? Leave it alone but with old data?

nsheaps replied:
I'd opt to leave it alone but make that error loud A F. We don't want to bring down a service because someone accidentally deleted something, but we also don't want that to go silently unnoticed.

TylerLubeck replied:
oh dang we handled this case in the previous form and I accidentally ripped it out, good catch

nsheaps commented on tests/unit/sidecars/live_data_watcher_loader_tests.py:
```diff
@@ -0,0 +1,104 @@
+
+@pytest.mark.parametrize('data,return_value', (
+    # Non-RAW configs
+    (json.dumps({'live_data_watcher_load_type': 'S3'}).encode('utf-8'), LoaderType.S3),
```
Should probably have a test for this saying something other than S3

KTAtkinson commented on tests/unit/sidecars/live_data_watcher_loader_tests.py:
```diff
@@ -0,0 +1,104 @@
+    (json.dumps({'key': 'value'}).encode('utf-8'), LoaderType.RAW),
+    (json.dumps([]).encode('utf-8'), LoaderType.RAW),
+    (json.dumps(1234).encode('utf-8'), LoaderType.RAW),
+    (json.dumps(None).encode('utf-8'), LoaderType.RAW),
```
What about the positive case? What if the contents don't parse as JSON at all?

nsheaps replied:
The positive test case is at the top. These are for negative parses, where it defaults back to RAW. Keep in mind that the data in S3 is not parsed, it's just dumped to a file.

TylerLubeck replied:
This is all testing parsing the znode data anyway. Positive case is at the top, still needs a 'non json' test case though. I'll add

KTAtkinson commented on baseplate/sidecars/live_data_watcher.py:
```diff
@@ -128,25 +73,67 @@ def on_change(self, data: Optional[bytes], _znode_stat: ZnodeStat) -> None:
+
+
+def _load_from_s3(data: bytes) -> Optional[bytes]:
+    loader_config = json.loads(data.decode("UTF-8"))
```
:no_bell: I think it's likely a good idea that we are representing this data as JSON and not ini, but using ini would allow us to use the built in config parsing spec. It might be a good idea to add a comment here about why we are using this json and not ini.

nsheaps replied:
🔕  Do we use ini config formats for anything other than loading ini files?

TylerLubeck replied:
Honestly I didn't even consider not using JSON for this, writing out ini from scratch in code is a PITA in my experience. I don't have much of a reason other than that, but wowsers I don't wanna change it either

KTAtkinson replied:
> no_bell Do we use ini config formats for anything other than loading ini files?

All application and sidecar configuration is in ini, and the [baseplate config library](https://github.com/reddit/baseplate.py/blob/develop/baseplate/lib/config.py) allows you to define a spec for your configuration that would take care of the validation below. I'm not aware of anyone using this outside of Baseplate application and sidecar configuration though,

KTAtkinson replied:
> Honestly I didn't even consider not using JSON for this, writing out ini from scratch in code is a PITA in my experience. I don't have much of a reason other than that, but wowsers I don't wanna change it either

To be clear I don't should use ini. It might be worth pointing out that we are doing some of the things that the baseplate config library does. To me the most compelling reason to use json, and not ini is it's simpler. It allows someone who is writing data to zk to verify their data a lot easier.

KTAtkinson commented on tests/unit/sidecars/live_data_watcher_loader_tests.py:
```diff
@@ -0,0 +1,104 @@
+    ClientError({'Error': {'Code': 'error', 'Message': 'msg'}}, 'operation'),
+))
+def test_unsuccessful_load_from_s3(start_mock_s3, exc):
+    with mock.patch('baseplate.sidecars.live_data_watcher.boto3.client') as client:
```
This use of a mock.MagicMock limits what you're able to test here. Did you consider using [`botocore.stub.Stubber`](https://botocore.amazonaws.com/v1/documentation/api/latest/reference/stubber.html) instead?

TylerLubeck replied:
TIL about Stubber, I was looking at moto and couldn't find anything in there so fell back to mock

nsheaps commented on baseplate/sidecars/live_data_watcher.py:
```diff
@@ -128,25 +73,67 @@ def on_change(self, data: Optional[bytes], _znode_stat: ZnodeStat) -> None:
+            "Failed to update live config: params for loading from S3 are incorrect. Received error: %s",
+            error,
+        )
+    return None
```
I would argue that if it failed to load data, it should return an exception, not None, which is a valid piece of data.

TylerLubeck committed:
Resolve simpler PR feedback

TylerLubeck committed:
Future proof loader s3 call checks

TylerLubeck committed:
Test and handle S3/boto errors

TylerLubeck committed:
run formatter

TylerLubeck committed:
Linting and typing

TylerLubeck commented on baseplate/sidecars/live_data_watcher.py:
```diff
@@ -20,93 +25,36 @@
+    S3 = "S3"
+
+    @classmethod
+    def _missing_(cls, value: object) -> "LoaderType":
```
I don't like this type [but](https://github.com/python/typeshed/blob/master/stdlib/enum.pyi#L176)

TylerLubeck commented on baseplate/sidecars/live_data_watcher.py:
```diff
@@ -128,25 +90,76 @@ def on_change(self, data: Optional[bytes], _znode_stat: ZnodeStat) -> None:
+def _parse_loader_type(data: bytes) -> LoaderType:
+    try:
+        json_data = json.loads(data.decode("UTF-8"))
+    except (UnicodeDecodeError, json.decoder.JSONDecodeError):
```
I ran the following against a copy of our zk instance and found `UnicodeDecodeError`, otherwise we're clean to load everything in there (even the things we don't actually intend to load)
 
```py
zk = kazooClient('ourcluster')
def parse_recursive(start='/'):
     print(f'Checking {start}')
     children  = zk.get_children(start)
     for c in children:
         parse_recursive(start + '/' + c)
     data, _ = zk.get(start)
     try:
         _parse_loader_type(data)
     except Exception as e:
         print(f'FAIL FAIL FAIL FAIL FAIL {e.__class__.__file__}: {str(e)} \n {data} \n')
         
   parse_recursive()
   ```

TylerLubeck committed:
Test cleanup

TylerLubeck committed:
more linting

TylerLubeck commented on tests/unit/sidecars/live_data_watcher_loader_tests.py:
```diff
@@ -0,0 +1,125 @@
+        assert _load_from_s3(json.dumps(data).encode("utf-8"))
+
+
+def test_successful_load_from_s3(start_mock_s3):
```
I suppose this could also be converted to Stubber from moto, @KTAtkinson do you have strong feelings?

KTAtkinson replied:
II would prefer using `Stubber` because it will catch misuses of the API. The way it works now, it assumes that any method or attribute you use on the client or it's return values is valid.

TylerLubeck replied:
Alright, stubber is in place but we still need some mocking.

Because we create a client based on the znode data, we have to create it at runtime. because we create it at runtime, we can't inject our stubbed session/client easily. So I mock out the client create call to always return our stubbed session/client and then we can use stubber everywhere

KTAtkinson commented on baseplate/sidecars/live_data_watcher.py:
```diff
@@ -128,25 +90,76 @@ def on_change(self, data: Optional[bytes], _znode_stat: ZnodeStat) -> None:
+    try:
+        json_data = json.loads(data.decode("UTF-8"))
+    except (UnicodeDecodeError, json.decoder.JSONDecodeError):
+        logger.debug("Data is not parseable as JSON, loading as PASSTHROUGH")
```
Do we know how often this log message might happen?

nsheaps replied:
![Screen Shot 2022-06-02 at 11 13 16 AM](https://user-images.githubusercontent.com/1282393/171662219-3b7d67d0-b18a-42f7-b37a-134cd4c60dea.png)

Basically if the zk node is empty or contains a string that isn't wrapped in quotes

KTAtkinson commented:
I'm putting this proposal here for lack of a better place to put it. I have an alternate approach that might solve a few of the unknowns here.

What if all nodes whose data we expect to exist in S3 had node names in Zookeeper with an agreed upon prefix or suffix. This would allow us to test to make sure all nodes conform to the configuration spec. It would also reduce the surface area for errors. It would require that we update the Zookeeper node name in configuration files.which is not ideal, but that's easily done with a batch change or other mechanism.

The big unknown for me in this case is that I'm not sure how zk permission work.

nsheaps commented:
@KTAtkinson I disagree with the approach you mentioned with ZK data that's named specific for referring to where the data actually lives, in part due to the fact that this is being added to provide some flexibility in where the data is being stored as it is expanding in size as time goes on. Keeping the location specific allows the owner of the data to change it's location without having to update services to point to the new ZK node.

nsheaps commented on baseplate/sidecars/live_data_watcher.py:
```diff
@@ -120,6 +68,20 @@ def on_change(self, data: Optional[bytes], _znode_stat: ZnodeStat) -> None:
+            logger.warning(
+                "No data to write to destination file. Something is likely misconfigured."
+            )
+            return
```
would prematurely returning here cause an application to potentially fail if the file doesn't exist? Should it be an error (maybe tracked with a span/metric of some sort that can alert on an error, sentry would suffice if it's monitored) and self-correct to an empty string to write?

TylerLubeck replied:
Potentially, but this matches behavior from before any of our S3 work: https://github.com/reddit/baseplate.py/blob/15b60b7f1100a1942299308e698b3db8993cd11e/baseplate/sidecars/live_data_watcher.py#L43

TylerLubeck replied:
That is to say, I'm more worried about the opposite - are there use cases where we're relying on the fact that an empty znode creates no file?

nsheaps commented on baseplate/sidecars/live_data_watcher.py:
```diff
@@ -128,25 +90,76 @@ def on_change(self, data: Optional[bytes], _znode_stat: ZnodeStat) -> None:
+    try:
+        json_data = json.loads(data.decode("UTF-8"))
+    except (UnicodeDecodeError, json.decoder.JSONDecodeError):
+        logger.debug("Data is not parseable as JSON, loading as PASSTHROUGH")
```
These are two different failure modes, feels like they should be split. When would the data not be utf-8 encoded?

TylerLubeck replied:
Only when it's binary data or something else shoved in to zookeeper. I found this one while running this method against all of our znodes - it only triggered for some znodes managed by other sources (flink, I _think_, but I don't remember) so we're unlikely to run in to it in our day-to-day, but I wanted to include it anyway because we've said that before

nsheaps commented on baseplate/sidecars/live_data_watcher.py:
```diff
@@ -120,6 +68,20 @@ def on_change(self, data: Optional[bytes], _znode_stat: ZnodeStat) -> None:
             self.handle_empty_data()
             return
 
+        loader_type = _parse_loader_type(data)
```
🔕  this may be better formatted as returning a tuple with the first element being the loader type and the second being the location to fetch from, if it's null then zk contains the data, otherwise use the return value in the _load_from_s3 call so you're not parsing `data` twice.

TylerLubeck replied:
IMO that makes the code dirtier and is kind of unnecessary. 
- We know the biggest blob will be 1mb
- Most of our znodes are far far less than 1mb
- The goal of this is that anything approaching 1mb will be converted to s3 and thus the znode drops back to <<<<1mb
- This only triggers on startup or when znodes change, so not that often.

there's perhaps an argument about the startup cost of parsing 1mb of json but I don't have the napkin math on what that'd be.

If we're really worried about it, I'd prefer we go one of two routes
- use ujson or orjson or similar
- Bail on the enums and build out a class tree of loaders. This is a lot for 2 loaders though.

TylerLubeck committed:
Add explanatory comments

TylerLubeck committed:
Use stubber for all s3 tests

We still have to mock the boto3 client object to make sure we inject
our stubbed client

TylerLubeck committed:
linting
